src/core/basics.ml

```diff
@@ -204,41 +213,36 @@ let occurs : meta -> term -> bool =
   try iter_meta false fn t; false with Found -> true
 
 (** [get_metas b t] returns the list of all the metavariables in [t], and in
-   the types of metavariables recursively if [b], sorted wrt [cmp_meta]. *)
+    the types of metavariables recursively if [b], sorted wrt [cmp_meta]. *)
 let get_metas : bool -> term -> meta list = fun b t ->
   let open Stdlib in
   let l = ref [] in
   iter_meta b (fun m -> l := m :: !l) t;
   List.sort_uniq cmp_meta !l
 
 (** [has_metas b t] checks whether there are metavariables in [t], and in the
-   types of metavariables recursively if [b] is true. *)
+    types of metavariables recursively if [b] is true. *)
 let has_metas : bool -> term -> bool =
   let exception Found in fun b t ->
   try iter_meta b (fun _ -> raise Found) t; false with Found -> true
 
-(** [distinct_vars_opt ts] checks that [ts] is made of distinct
-   variables and returns these variables. *)
-let distinct_vars_opt : term array -> tvar array option =
-  let exception Not_unique_var in fun ts ->
-  let open Stdlib in
-  let vars = ref VarSet.empty in
-  let to_var t =
-    match unfold t with
-    | Vari x when not (VarSet.mem x !vars) -> vars := VarSet.add x !vars; x
-    | _ -> raise Not_unique_var
-  in try Some (Array.map to_var ts) with Not_unique_var -> None
-
-(** [distinct_vars ts] checks that [ts] is made of distinct variables. *)
-let distinct_vars : term array -> bool =
-  let exception Not_unique_var in fun ts ->
+(** [distinct_vars ctx ts]  checks  that terms of  [ts] are made of  variables
+    that are themselves or their definition in  [ctx] (if it exists) distinct.
+    If so, the variables are returned. *)
+let distinct_vars : ctxt -> term array -> tvar array option = fun ctx ts ->
+  let exception Not_unique_var in
   let open Stdlib in
   let vars = ref VarSet.empty in
-  let check t =
+  let rec to_var t =
     match unfold t with
```

Comment:
I would prefer to introduce a function unfold_ctxt in ctxt.ml similar to unfold but with some new case:
```
let unfold_ctxt : ctxt -> term -> term = fun c t ->
  let rec unfold t =
  match t with
  | Meta(m, ar)          ->
      begin
        match !(m.meta_value) with
        | None    -> t
        | Some(b) -> unfold (Bindlib.msubst b ar)
      end
  | TEnv(TE_Some(b), ar) -> unfold (Bindlib.msubst b ar)
  | TRef(r)              ->
      begin
        match !r with
        | None    -> t
        | Some(v) -> unfold v
      end
| Vari(x) ->
  begin
    match def_of x ctx with
    | None -> t
    | Some u -> unfold u
    end
  | _                    -> t
  in unfold
```
Then, distinct_vars and nl_distinct_vars (that could be moved to basics as well) could be defined as follows:
```
let to_var t = match unfold_ctxt ctx t with Vari x ... | _ -> raise ... in
```
This function unfold_ctxt may have other applications later.
In fact, we could redefine Basics.unfold as `unfold_ctxt []`. unfold_ctxt could then be moved to terms.ml if def_of was moved there too.

Reply:
I moved `def_of` and `unfold_ctx` to `terms.ml`

Reply:
Then, you should move type_of as well.